eip-4626.md

```diff
@@ -0,0 +1,152 @@
+
+A standard for yield bearing vaults will allow for a similar cambrian explosion to ERC-20, unlocking access to yield in a variety of applications with little specialized effort from developers.
+
+
+## Specification
+
+### Methods
+
+#### deposit
+
+`function deposit(address _to, uint256 _value) public returns (uint256 _shares)`
+
+Deposits `_value` tokens into the vault and grants ownership of them to `_to`.
+
+MAY return a pro-rata ownership `_shares` value corresponding `_value`, if not MUST return `0`.
+
+MAY represent `_shares` using internal accounting or an ERC-20 token. 
+
+If pro-rata shares ownership is implemented, the vault SHOULD implement `balanceOf`, `redeem`, `totalSupply` and `exchangeRate`.
+
+
+#### withdraw
+
+`function withdraw(address _to, uint256 _value) public returns (uint256 _shares)`
+
+Withdraws `_value` tokens from the vault and transfers them to `_to`. 
+
+MAY return a pro-rata ownership `_shares` value corresponding to `_value`,  if not MUST return `0`.
+
+
+#### totalHoldings
+`function totalHoldings() public view returns (uint256)`
+
+Returns the total amount of underlying tokens held/managed by the vault.
+
+
+#### balanceOfUnderlying
+`function balanceOfUnderlying(address _owner) public view returns (uint256)`
+
+Returns the total amount underlying tokens held in the vault for `_owner`.
+
+#### underlying
+`function underlying() public view returns (address)`
+
+Returns the address of the token the vault uses for accounting, depositing, and withdrawing. 
+
+SHOULD return a token implementing the ERC-20 standard.
+
+
+#### totalSupply
+
+`function totalSupply() public view returns (uint256)`
+
+Returns the total number of unredeemed vault shares in circulation.
+
+OPTIONAL - This method is only needed for vaults that implement a pro-rata share mechanism for deposits.
+
+#### balanceOf
+
+`function balanceOf(address _owner) public view returns (uint256)`
+
+Returns the total amount of vault shares the `_owner` currently has.
+
+OPTIONAL - This method is only needed for vaults that implement a pro-rata share mechanism for deposits.
+
+#### redeem
+
+`function redeem(address _to, uint256 _shares) public returns (uint256 _value)`
+
+Redeems a specific number of `_shares` for underlying tokens and transfers them to `_to`. 
+
+MAY return a pro-rata ownership `_shares` value corresponding `_value`, if not MUST return `0`.
+
+OPTIONAL - This method is only needed for vaults that implement a pro-rata share mechanism for deposits.
+
+#### exchangeRate
+
+`function exchangeRate() public view returns (uint256)`
+
+The amount of underlying tokens one `baseUnit` of vault shares is redeemable for. 
+
+e.g. `_shares * exchangeRate() / baseUnit() = _value`.
+
+`exchangeRate() * totalSupply()` MUST equal `totalHoldings()`.
```

Comment:
we need to divide by the right scalar cuz totalHoldings decimals is scaled by underlying decimals

Reply:
Great call, should we remove this MUST condition or add the decimals scalar? Leaning former

Reply:
im not too opinionated either way tbh, just want to make sure we clarify the decimals one way or another

Reply:
It's easier and more general to use a `calculateShares(uint256 underlyingAmount) returns (uint256 shares amount)`. No need to think about scalars, and supports non-linear exchange rates with precision.

Likewise, you should have the symmetrical `calculateUnderlying(uint256 sharesAmount) returns (uint256 underlyingAmount)`.

This pattern has been used in the yield oracles with great success, and partially in the lido staked ether contract where you have also symmetrical functions for asset conversions.

Reply:
To give another example, if a vault is implemented with a fee on deposits or withdrawals, it won't be possible for an interesting contract to know the deposit or withdrawal output, without specific knowledge of those fees.

By putting any conversion logic behind the `calculate` functions, vaults with non-linear math or fees can be integrated seamlessly.

Reply:
I like this suggestion

Reply:
Id say in addition to. 

Like @alcueca's saying, a `calculateX` is needed to ensure your own internal bookkeeping is in line for a single specific `deposit`/`withdraw` that could include unknown fees, but for more generalized bookkeeping that isnt tied to a single tx, it's still ideal to record the `exchangeRate`.

e.g. In our contracts, we benchmark `exchangeRate` at maturity (potentially regardless of any deposits/withdrawals) and use this as a point to begin divesting any additional yield from YT holders, back to PT holders.

Reply:
I would suggest this as instead `exchangeRate`. From the point of an integrating contract, you are not usually interested in the exchange rate itself, but in a specific conversion of a for b.

If you would still be interested in the exchange rate as an integrating contract, you can obtain it by passing one full unit as the amount parameter.

Furthermore, it removes the constraint on how tokens are converted for shares, which the EIP forces to a linear, no-fee formula (otherwise `exchangeRate` is true only for one full unit).

If still unconvinced, I'd like to hear the arguments in favour of a `exchangeRate` variable over two symmetrical functions.

Reply:
I'm writing a generic ERC4626 router and all of your advice has become very appropriate as I encounter some of the nuances of the implementation details between vaults of different types